src/libraries/System.Private.CoreLib/src/System/SpanHelpers.cs

```diff
@@ -403,5 +406,222 @@ public static unsafe void ClearWithReferences(ref IntPtr ip, nuint pointerSizeLe
+                    Unsafe.As<byte, Vector128<int>>(ref last) = tempFirst;
+
+                    // Adjust the references to point to the next vector
+                    first = ref Unsafe.Add(ref first, Vector128<byte>.Count);
+                    last = ref Unsafe.Subtract(ref last, Vector128<byte>.Count);
+                    numBytesWritten += Vector128<byte>.Count * 2;
+                } while ((numBytes - numBytesWritten) >= Vector128<byte>.Count * 2);
+            }
+
+            // Store any remaining values one-by-one
+            last = ref Unsafe.Add(ref first, (numBytes - numBytesWritten) - sizeof(int));
+            ref int firstInt = ref Unsafe.As<byte, int>(ref first);
+            ref int lastInt = ref Unsafe.As<byte, int>(ref last);
+            while (numBytes - numBytesWritten > sizeof(int))
+            {
+                (lastInt, firstInt) = (firstInt, lastInt);
+                firstInt = ref Unsafe.Add(ref firstInt, 1);
+                lastInt = ref Unsafe.Subtract(ref lastInt, 1);
+                numBytesWritten += sizeof(int) * 2;
+            }
+        }
+
+        public static void Reverse(ref long buf, nuint length)
+        {
+            nint numBytes = (int)length * sizeof(long);
+            int numBytesWritten = 0;
+            ref byte first = ref Unsafe.As<long, byte>(ref buf);
+            ref byte last = ref Unsafe.Subtract(ref Unsafe.Add(ref first, numBytes), sizeof(long));
+            if (Avx2.IsSupported && Vector256<byte>.Count * 2 <= numBytes)
+            {
+                last = ref Unsafe.Subtract(ref Unsafe.Add(ref first, numBytes), Vector256<byte>.Count);
+                do
+                {
+                    // Load in values from beginning and end of the array.
+                    Vector256<int> tempFirst = Unsafe.As<byte, Vector256<int>>(ref first);
+                    Vector256<int> tempLast = Unsafe.As<byte, Vector256<int>>(ref last);
+
+                    // Avx2 operates on two 128-bit lanes rather than the full 256-bit vector.
+                    // Perform a shuffle to reverse each 128-bit lane, then permute to finish reversing the vector:
+                    //     +---------------+
+                    //     | A | B | C | D |
+                    //     +---------------+
+                    //         Shuffle --->
+                    //     +---------------+
+                    //     | B | A | D | C |
+                    //     +---------------+
+                    //         Permute --->
+                    //     +---------------+
+                    //     | D | C | B | A |
+                    //     +---------------+
+                    tempFirst = Avx2.Shuffle(tempFirst, 0b01001110);
+                    tempFirst = Avx2.Permute2x128(tempFirst, tempFirst, 1);
+                    tempLast = Avx2.Shuffle(tempLast, 0b01001110);
+                    tempLast = Avx2.Permute2x128(tempLast, tempLast, 1);
```

Comment:
These could use permute4x64 rather than shuffle+permute2x128.

For that matter, the `int` version could use permute8x32.  That would re-introduce the need for a memory load for the mask but on Skylake hardware, both the shuffles and permutes all get dispatched to a single port, so it may be a perf win to eliminate that bottleneck.

Reply:
Agreed that this is likely a perf win, especially for larger buffers. Updated.